File: Android/app/src/main/java/com/marine/fishtank/model/DataSource.kt
package com.marine.fishtank.model

/**
 * Created by MarineSeo on 2021-06-02.
 */
enum class Status {
    SUCCESS,
    ERROR,
    LOADING
}

data class DataSource<out T>(val status: Status, val data: T ) {
    companion object {
        fun <T> success(data: T): DataSource<T> {
            return DataSource(Status.SUCCESS, data)
        }

        fun <T> error(data: T): DataSource<T> {
            return DataSource(Status.ERROR, data)
        }

        fun <T> loading(data: T): DataSource<T> {
            return DataSource(Status.LOADING, data)
        }
    }
}

File: FishTankFlutter/lib/model/data_source.dart

enum Status { none, loading, complete, error }

class DataSource<T> {
  Status? status;
  T? data;
  String? message;

  DataSource(this.status, this.data);

  DataSource.none(): status = Status.none;

  DataSource.loading(): status = Status.loading;

  DataSource.complete(this.data): status = Status.complete;

  DataSource.error(this.message): status = Status.error;
}